File: Shield.rb
require_relative 'CombatElement'
module Irrgarten

  # La clase Shield representa un escudo utilizado en combate para proteger al jugador.
  #
  # Esta clase administra el nivel de protección que aporta el escudo y la cantidad de usos restantes.
  # Además, provee la funcionalidad para ejecutar la acción de descarte (discard) a través de la clase Dice.
  #
  # @author Juan Caballero Santoyo
  class Shield < CombatElement


    # Realiza una acción de protección con el escudo.
    #
    # Si el escudo cuenta con usos disponibles, se decrementa en uno el contador de usos y se retorna el valor
    # de protección asociado; de lo contrario, se retorna 0, indicando que no se puede usar el escudo.
    #
    # @return [Integer] El valor de protección aplicado o 0 si no hay usos restantes.
    def protect
      self.produce_effect
    end


    # Retorna una representación en cadena del escudo.
    #
    # El formato de la cadena es "W[<protección>, <usos>]", donde se muestran el valor actual de protección
    # y la cantidad de usos restantes.
    #
    # @return [String] La representación textual del escudo.
    def to_s
      return "S"+super
    end
    
    public_class_method :new
  end

end

File: java/src/irrgarten/Shield.java

package irrgarten;

/**
 * Clase que representa los escudos que utiliza el jugador cuando se defiende
 * de un ataque de un monstruo.
 * 
 * @author Juan Caballero Santoyo
 */
public class Shield extends CombatElement {
    
    /**
     * Constructor de la clase Shield
     * @param protection El número que indica la protección que aporta el escudo
     * @param uses El número de usos disponibles del elemento de combate
     */
    public Shield(float protection, int uses){
        super(protection,uses);
    }
    
    /**
     * Método para realizar una protección con el escudo.
     * @return La intensidad de la protección del escudo.
     */
    public float protect(){
        return this.produceEffect();
    }
     
    /**
     * Método que devuelve una representación en forma de cadena de caracteres
     * del estado interno del objeto.
     * @return Representación en forma de cadena de caracteres del estado interno del objeto.
     */
    @Override
    public String toString(){
        return "S" + super.toString();
    }
   
}
